File: apps/api/plane/api/serializers/workflow.py
# Third party imports
from rest_framework import serializers

# Module imports
from .base import BaseSerializer
from .state import StateLiteSerializer
from .user import UserLiteSerializer
from plane.db.models import (
    WorkflowTemplate,
    WorkflowState,
    WorkflowTransition,
    WorkflowTransitionReviewer,
    WorkflowAssignmentRule,
    WorkflowTransitionLog,
)


class WorkflowTemplateSerializer(BaseSerializer):
    created_by_detail = UserLiteSerializer(source="created_by", read_only=True)
    total_states = serializers.SerializerMethodField()
    total_transitions = serializers.SerializerMethodField()
    
    class Meta:
        model = WorkflowTemplate
        fields = [
            "id",
            "name",
            "description",
            "is_active",
            "is_default",
            "created_by",
            "created_by_detail",
            "created_at",
            "updated_at",
            "total_states",
            "total_transitions",
        ]
        read_only_fields = [
            "id",
            "created_by",
            "created_at",
            "updated_at",
            "workspace",
            "project",
            "total_states",
            "total_transitions",
        ]

    def get_total_states(self, obj):
        return obj.workflow_states.count()

    def get_total_transitions(self, obj):
        return obj.workflow_transitions.count()

    def validate(self, data):
        # If the default is being provided then make all other workflows default False
        if data.get("is_default", False):
            WorkflowTemplate.objects.filter(
                project_id=self.context.get("project_id")
            ).update(is_default=False)
        return data


class WorkflowStateSerializer(BaseSerializer):
    state_detail = StateLiteSerializer(source="state", read_only=True)
    
    class Meta:
        model = WorkflowState
        fields = [
            "id",
            "workflow",
            "state",
            "state_detail",
            "sequence",
            "allow_new_issues",
            "conditions",
            "created_at",
            "updated_at",
        ]
        read_only_fields = [
            "id",
            "workflow",
            "created_at",
            "updated_at",
            "workspace",
            "project",
        ]


class WorkflowTransitionReviewerSerializer(BaseSerializer):
    reviewer_detail = UserLiteSerializer(source="reviewer", read_only=True)
    
    class Meta:
        model = WorkflowTransitionReviewer
        fields = [
            "id",
            "transition",
            "reviewer",
            "reviewer_detail",
            "created_at",
        ]
        read_only_fields = [
            "id",
            "created_at",
            "updated_at",
            "workspace",
            "project",
        ]


class WorkflowTransitionSerializer(BaseSerializer):
    from_state_detail = StateLiteSerializer(source="from_state", read_only=True)
    to_state_detail = StateLiteSerializer(source="to_state", read_only=True)
    reviewers = WorkflowTransitionReviewerSerializer(many=True, read_only=True)
    
    class Meta:
        model = WorkflowTransition
        fields = [
            "id",
            "workflow",
            "from_state",
            "from_state_detail",
            "to_state",
            "to_state_detail",
            "conditions",
            "actions",
            "require_reviewer",
            "reviewers",
            "created_at",
            "updated_at",
        ]
        read_only_fields = [
            "id",
            "workflow",
            "created_at",
            "updated_at",
            "workspace",
            "project",
        ]


class WorkflowAssignmentRuleSerializer(BaseSerializer):
    class Meta:
        model = WorkflowAssignmentRule
        fields = [
            "id",
            "workflow",
            "condition_field",
            "condition_value",
            "priority",
            "is_active",
            "created_at",
            "updated_at",
        ]
        read_only_fields = [
            "id",
            "created_at",
            "updated_at",
            "workspace",
            "project",
        ]


class WorkflowTransitionLogSerializer(BaseSerializer):
    from_state_detail = StateLiteSerializer(source="from_state", read_only=True)
    to_state_detail = StateLiteSerializer(source="to_state", read_only=True)
    reviewer_detail = UserLiteSerializer(source="reviewer", read_only=True)
    actor_detail = UserLiteSerializer(source="actor", read_only=True)
    
    class Meta:
        model = WorkflowTransitionLog
        fields = [
            "id",
            "issue",
            "workflow",
            "transition",
            "from_state",
            "from_state_detail",
            "to_state",
            "to_state_detail",
            "reviewer",
            "reviewer_detail",
            "actor",
            "actor_detail",
            "comment",
            "metadata",
            "created_at",
        ]
        read_only_fields = [
            "id",
            "created_at",
            "updated_at",
            "workspace",
            "project",
        ]


# Detailed serializers for workflow management
class WorkflowTemplateDetailSerializer(WorkflowTemplateSerializer):
    workflow_states = WorkflowStateSerializer(many=True, read_only=True)
    workflow_transitions = WorkflowTransitionSerializer(many=True, read_only=True)
    assignment_rules = WorkflowAssignmentRuleSerializer(many=True, read_only=True)
    
    class Meta(WorkflowTemplateSerializer.Meta):
        fields = WorkflowTemplateSerializer.Meta.fields + [
            "workflow_states",
            "workflow_transitions", 
            "assignment_rules",
        ]


class WorkflowValidationSerializer(BaseSerializer):
    """Serializer for validating workflow state transitions"""
    from_state_id = serializers.UUIDField()
    to_state_id = serializers.UUIDField()
    issue_id = serializers.UUIDField()
    comment = serializers.CharField(required=False, allow_blank=True)
    
    class Meta:
        model = WorkflowTransitionLog
        fields = [
            "from_state_id",
            "to_state_id", 
            "issue_id",
            "comment",
        ]

File: packages/types/src/workflow.ts
export type IWorkflowTemplate = {
  id: string;
  name: string;
  description?: string;
  is_active: boolean;
  is_default: boolean;
  created_by: string;
  created_by_detail?: {
    id: string;
    display_name: string;
    avatar?: string;
    first_name: string;
    last_name: string;
  };
  created_at: string;
  updated_at: string;
  total_states?: number;
  total_transitions?: number;
  workflow_states?: IWorkflowState[];
  workflow_transitions?: IWorkflowTransition[];
  assignment_rules?: IWorkflowAssignmentRule[];
};

export type IWorkflowState = {
  id: string;
  workflow: string;
  state: string;
  state_detail?: {
    id: string;
    name: string;
    color: string;
    group: string;
  };
  sequence: number;
  allow_new_issues: boolean;
  conditions?: Record<string, any>;
  created_at: string;
  updated_at: string;
};

export type IWorkflowTransition = {
  id: string;
  workflow: string;
  from_state: string;
  from_state_detail?: {
    id: string;
    name: string;
    color: string;
    group: string;
  };
  to_state: string;
  to_state_detail?: {
    id: string;
    name: string;
    color: string;
    group: string;
  };
  conditions?: Record<string, any>;
  actions?: Record<string, any>;
  require_reviewer: boolean;
  reviewers?: IWorkflowTransitionReviewer[];
  created_at: string;
  updated_at: string;
};

export type IWorkflowTransitionReviewer = {
  id: string;
  transition: string;
  reviewer: string;
  reviewer_detail?: {
    id: string;
    display_name: string;
    avatar?: string;
    first_name: string;
    last_name: string;
  };
  created_at: string;
};

export type IWorkflowAssignmentRule = {
  id: string;
  workflow: string;
  condition_field: "label" | "assignee" | "priority" | "custom_field" | "issue_type";
  condition_value: any;
  priority: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
};

export type IWorkflowTransitionLog = {
  id: string;
  issue: string;
  workflow: string;
  transition?: string;
  from_state: string;
  from_state_detail?: {
    id: string;
    name: string;
    color: string;
    group: string;
  };
  to_state: string;
  to_state_detail?: {
    id: string;
    name: string;
    color: string;
    group: string;
  };
  reviewer?: string;
  reviewer_detail?: {
    id: string;
    display_name: string;
    avatar?: string;
    first_name: string;
    last_name: string;
  };
  actor: string;
  actor_detail?: {
    id: string;
    display_name: string;
    avatar?: string;
    first_name: string;
    last_name: string;
  };
  comment?: string;
  metadata?: Record<string, any>;
  created_at: string;
};

export type IWorkflowValidation = {
  from_state_id: string;
  to_state_id: string;
  issue_id: string;
  comment?: string;
};

export type IWorkflowValidationResponse = {
  allowed: boolean;
  reason?: string;
  requires_reviewer?: boolean;
  reviewers?: string[];
  transition_id?: string;
};

// Form types for creating/updating workflows
export type IWorkflowTemplateFormData = {
  name: string;
  description?: string;
  is_active?: boolean;
  is_default?: boolean;
};

export type IWorkflowStateFormData = {
  state: string;
  sequence: number;
  allow_new_issues?: boolean;
  conditions?: Record<string, any>;
};

export type IWorkflowTransitionFormData = {
  from_state: string;
  to_state: string;
  conditions?: Record<string, any>;
  actions?: Record<string, any>;
  require_reviewer?: boolean;
  reviewer_ids?: string[];
};

export type IWorkflowAssignmentRuleFormData = {
  condition_field: "label" | "assignee" | "priority" | "custom_field" | "issue_type";
  condition_value: any;
  priority?: number;
  is_active?: boolean;
};
